di-ipv-credential-issuer-stub/src/main/java/uk/gov/di/ipv/stub/cred/auth/ClientJwtVerifier.java

```diff
@@ -0,0 +1,71 @@
+package uk.gov.di.ipv.stub.cred.auth;
+
+import com.nimbusds.jose.JOSEException;
+import com.nimbusds.oauth2.sdk.ParseException;
+import com.nimbusds.oauth2.sdk.auth.PrivateKeyJWT;
+import com.nimbusds.oauth2.sdk.auth.verifier.ClientAuthenticationVerifier;
+import com.nimbusds.oauth2.sdk.auth.verifier.InvalidClientException;
+import com.nimbusds.oauth2.sdk.id.Audience;
+import spark.QueryParamsMap;
+import uk.gov.di.ipv.stub.cred.config.ClientConfig;
+import uk.gov.di.ipv.stub.cred.config.CredentialIssuerConfig;
+import uk.gov.di.ipv.stub.cred.error.ClientAuthenticationException;
+
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+public class ClientJwtVerifier {
+
+    public static final String AUTHENTICATION_METHOD = "authenticationMethod";
+    public static final String NONE = "none";
+
+    private final ClientAuthenticationVerifier<Object> verifier;
+
+    public ClientJwtVerifier() {
+        this.verifier = getPopulatedClientAuthVerifier();
+    }
+
+    public void authenticateClient(QueryParamsMap requestParams)
+            throws ClientAuthenticationException {
+
+        PrivateKeyJWT authenticationJwt;
+        try {
+            authenticationJwt = PrivateKeyJWT.parse(listifyParamValues(requestParams));
+        } catch (ParseException e) {
+            throw new ClientAuthenticationException(e);
+        }
+
+        ClientConfig clientConfig = CredentialIssuerConfig.getClientConfig(authenticationJwt.getClientID().getValue());
+        if (clientConfig == null) {
+            throw new ClientAuthenticationException(String.format("Config for client ID '%s' not found", authenticationJwt.getClientID().getValue()));
+        }
+
+        if (clientConfig.getJwtAuthentication().get(AUTHENTICATION_METHOD).equals(NONE)) {
+            return;
+        }
+
+        try {
+            verifier.verify(authenticationJwt, null, null);
+        } catch (InvalidClientException | JOSEException e) {
+            throw new ClientAuthenticationException(e);
+        }
+    }
+
+    private Map<String, List<String>> listifyParamValues(QueryParamsMap requestParams) {
```

Comment:
We might not need this method. The PrivateKeyJWT.parse(queryParams) has a few different overloaded methods. One of them accepts just a single string value which are the query params i.e "var1=hello&var2=bye&var=hello" etc...

So maybe instead of passing in the QueryMap object into the authenticateClient method, we can instead pass in the original query params as a string, which I think we can get from something like

request.queryString()  #this would be from within the TokenHandler where the authenticateClient method is called from

Reply:
Nice. Have pushed a commit to use the query string version instead.